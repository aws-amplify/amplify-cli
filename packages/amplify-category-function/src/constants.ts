export const category = 'function';

export const envVarPrintoutPrefix =
  '\nYou can access the following resource attributes as environment variables from your Lambda function\n\t';
const topLevelCommentBlockTitle = 'Amplify Params - DO NOT EDIT';
export const topLevelCommentPrefix = `/* ${topLevelCommentBlockTitle}\n\t`;
export const topLevelCommentSuffix = `\n${topLevelCommentBlockTitle} */`;

export enum CRUDOperation {
  CREATE = 'create',
  READ = 'read',
  UPDATE = 'update',
  DELETE = 'delete',
}
