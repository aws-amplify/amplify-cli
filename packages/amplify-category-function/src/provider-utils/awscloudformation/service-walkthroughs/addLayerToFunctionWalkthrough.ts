import { $TSContext, stateManager } from 'amplify-cli-core';
import { FunctionDependency, FunctionParameters, FunctionRuntime, LambdaLayer } from 'amplify-function-plugin-interface';
import { askCustomArnQuestion, askLayerOrderQuestion, askLayerSelection } from '../utils/addLayerToFunctionUtils';

const confirmationPrompt = 'Do you want to enable Lambda layers for this function?';

/**
 * Performs the walkthrough to add layers to a function
 * @param context Amplify platform context
 * @param runtime Runtime of the function that is being modified
 * @param previousSelections Array of layers already added to this function (if any)
 * @param defaultConfirm Determines whether the confirmation question defaults to yes or no
 */
export const addLayersToFunctionWalkthrough = async (
  context: $TSContext,
  runtime: Pick<FunctionRuntime, 'value'>,
  previousSelections: LambdaLayer[] = [],
  defaultConfirm = false,
): Promise<Required<Pick<FunctionParameters, 'lambdaLayers' | 'dependsOn'>>> => {
  let lambdaLayers: LambdaLayer[] = [];
  let dependsOn: FunctionDependency[] = [];

  // ask initial confirmation
  if (!(await context.amplify.confirmPrompt(confirmationPrompt, defaultConfirm))) {
    return { lambdaLayers: previousSelections, dependsOn };
  }

  let askArnQuestion: boolean;
  ({ lambdaLayers, dependsOn, askArnQuestion } = await askLayerSelection(
    context,
    stateManager.getMeta(),
    runtime.value,
    previousSelections,
  ));

  if (askArnQuestion) {
    lambdaLayers = lambdaLayers.concat(await askCustomArnQuestion(lambdaLayers.length, previousSelections));
  }

  if (lambdaLayers.length === 0) {
    context.print.info('No Lambda layers were selected');
    if (previousSelections.length > 0) {
      const plural = previousSelections.length > 1 ? 's' : '';
      const removeMessage = `Removing ${previousSelections.length} previously added Lambda layer${plural} from Lambda function`;
      context.print.info(removeMessage);
    }
  }

  lambdaLayers = await askLayerOrderQuestion(lambdaLayers, previousSelections);

  return { lambdaLayers, dependsOn };
};
