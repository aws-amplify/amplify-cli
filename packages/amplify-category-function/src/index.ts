import path from 'path';
import { category } from './constants';
import { categoryName } from './provider-utils/awscloudformation/utils/constants';
import { FunctionBreadcrumbs, FunctionRuntimeLifecycleManager } from 'amplify-function-plugin-interface';
import sequential from 'promise-sequential';
import { updateConfigOnEnvInit } from './provider-utils/awscloudformation';

export async function add(context, providerName, service, parameters) {
  const options = {
    service,
    providerPlugin: providerName,
    build: true,
  };
  const providerController = require(`./provider-utils/${providerName}/index`);
  if (!providerController) {
    context.print.error('Provider not confgiured for this category');
    return;
  }
  return providerController.addResource(context, category, service, options, parameters);
}

export async function update(context, providerName, service, parameters, resourceToUpdate) {
  const providerController = require(`./provider-utils/${providerName}/index`);
  if (!providerController) {
    context.print.error('Provider not confgiured for this category');
    return;
  }
  return providerController.updateResource(context, category, service, parameters, resourceToUpdate);
}

export async function console(context) {
  context.print.info(`to be implemented: ${category} console`);
}

export async function migrate(context) {
  const { projectPath, amplifyMeta } = context.migrationInfo;
  const migrateResourcePromises = [];
  Object.keys(amplifyMeta).forEach(categoryName => {
    if (categoryName === category) {
      Object.keys(amplifyMeta[category]).forEach(resourceName => {
        try {
          const providerController = require(`./provider-utils/${amplifyMeta[category][resourceName].providerPlugin}/index`);
          if (providerController) {
            migrateResourcePromises.push(
              providerController.migrateResource(context, projectPath, amplifyMeta[category][resourceName].service, resourceName),
            );
          } else {
            context.print.error(`Provider not configured for ${category}: ${resourceName}`);
          }
        } catch (e) {
          context.print.warning(`Could not run migration for ${category}: ${resourceName}`);
          throw e;
        }
      });
    }
  });

  await Promise.all(migrateResourcePromises);
}

export async function getPermissionPolicies(context, resourceOpsMapping) {
  const amplifyMetaFilePath = context.amplify.pathManager.getAmplifyMetaFilePath();
  const amplifyMeta = context.amplify.readJsonFile(amplifyMetaFilePath);
  const permissionPolicies = [];
  const resourceAttributes = [];

  Object.keys(resourceOpsMapping).forEach(resourceName => {
    try {
      const providerController = require(`./provider-utils/${amplifyMeta[category][resourceName].providerPlugin}/index`);
      if (providerController) {
        const { policy, attributes } = providerController.getPermissionPolicies(
          context,
          amplifyMeta[category][resourceName].service,
          resourceName,
          resourceOpsMapping[resourceName],
        );
        permissionPolicies.push(policy);
        resourceAttributes.push({ resourceName, attributes, category });
      } else {
        context.print.error(`Provider not configured for ${category}: ${resourceName}`);
      }
    } catch (e) {
      context.print.warning(`Could not get policies for ${category}: ${resourceName}`);
      throw e;
    }
  });
  return { permissionPolicies, resourceAttributes };
}

export async function initEnv(context) {
  const { amplify } = context;
  const { resourcesToBeCreated, resourcesToBeDeleted, resourcesToBeUpdated } = await amplify.getResourceStatus('function');

  resourcesToBeDeleted.forEach(authResource => {
    amplify.removeResourceParameters(context, 'function', authResource.resourceName);
  });

  const tasks = resourcesToBeCreated.concat(resourcesToBeUpdated);

  const functionTasks = tasks.map(functionResource => {
    const { resourceName } = functionResource;
    return async () => {
      const config = await updateConfigOnEnvInit(context, 'function', resourceName);
      context.amplify.saveEnvResourceParameters(context, 'function', resourceName, config);
    };
  });

  await sequential(functionTasks);
}

// returns a function that can be used to invoke the lambda locally
export async function getInvoker(context: any, params: InvokerParameters): Promise<({ event: any }) => Promise<any>> {
  const resourcePath = path.join(context.amplify.pathManager.getBackendDirPath(), categoryName, params.resourceName);
  const breadcrumbs: FunctionBreadcrumbs = context.amplify.readBreadcrumbs(context, categoryName, params.resourceName);
  const runtimeManager: FunctionRuntimeLifecycleManager = await context.amplify.loadRuntimePlugin(context, breadcrumbs.pluginId);

  const lastBuildTimestampStr = (await context.amplify.getResourceStatus(category, params.resourceName)).allResources.find(
    resource => resource.resourceName === params.resourceName,
  ).lastBuildTimeStamp as string;

  return async request =>
    await runtimeManager.invoke({
      handler: params.handler,
      event: JSON.stringify(request.event),
      env: context.amplify.getEnvInfo().envName,
      runtime: breadcrumbs.functionRuntime,
      srcRoot: resourcePath,
      envVars: params.envVars,
      lastBuildTimestamp: new Date(lastBuildTimestampStr),
    });
}

export async function executeAmplifyCommand(context) {
  let commandPath = path.normalize(path.join(__dirname, 'commands'));
  if (context.input.command === 'help') {
    commandPath = path.join(commandPath, category);
  } else {
    commandPath = path.join(commandPath, category, context.input.command);
  }

  const commandModule = require(commandPath);
  await commandModule.run(context);
}

export async function handleAmplifyEvent(context, args) {
  context.print.info(`${category} handleAmplifyEvent to be implemented`);
  context.print.info(`Received event args ${args}`);
}

// Object used for internal invocation of lambda functions
export type InvokerParameters = {
  resourceName: string;
  handler: string;
  envVars?: { [key: string]: string };
};
