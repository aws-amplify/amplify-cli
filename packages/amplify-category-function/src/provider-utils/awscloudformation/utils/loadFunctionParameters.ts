import { $TSAny, JSONUtilities } from 'amplify-cli-core';
import _ from 'lodash';
import * as path from 'path';
import { functionParametersFileName } from './constants';

export const loadFunctionParameters = (resourcePath: string) => {
  const funcParams = JSONUtilities.readJson<$TSAny>(path.join(resourcePath, functionParametersFileName), { throwIfNotExist: false }) || {};

  // there was a bug where permissions were nested within "mutableParametersState" in the file so the following is necessary to ensure
  // forward compatability with functions whose permissions were updated with a version of the CLI where the bug existed
  if (funcParams.mutableParametersState) {
    _.assign(funcParams, { ...funcParams.mutableParametersState });
    delete funcParams.mutableParametersState;
  }
  return funcParams;
};
