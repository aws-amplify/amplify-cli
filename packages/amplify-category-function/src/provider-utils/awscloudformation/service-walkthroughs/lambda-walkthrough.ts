import { $TSAny, $TSContext, $TSObject, JSONUtilities, pathManager, stateManager } from 'amplify-cli-core';
import { FunctionParameters, ProjectLayer } from 'amplify-function-plugin-interface';
import inquirer from 'inquirer';
import _ from 'lodash';
import path from 'path';
import { categoryName, envVarPrintoutPrefix, topLevelCommentPrefix, topLevelCommentSuffix } from '../../../constants';
import { getNewCFNEnvVariables, getNewCFNParameters } from '../utils/cloudformationHelpers';
import {
  advancedSettingsList,
  cronJobSetting,
  functionParametersFileName,
  lambdaLayerSetting,
  secretsConfiguration,
  parametersFileName,
  resourceAccessSetting,
  ServiceName,
  environmentVariableSetting,
} from '../utils/constants';
import { merge } from '../utils/funcParamsUtils';
import { runtimeWalkthrough, templateWalkthrough } from '../utils/functionPluginLoader';
import { convertLambdaLayerMetaToLayerCFNArray } from '../utils/layerArnConverter';
import { loadFunctionParameters } from '../utils/loadFunctionParameters';
import { getStoredEnvironmentVariables } from '../utils/environmentVariablesHelper';
import { askEnvironmentVariableQuestions } from './environmentVariableWalkthrough';
import {
  fetchPermissionCategories,
  fetchPermissionResourcesForCategory,
  fetchPermissionsForResourceInCategory,
} from '../utils/permissionMapUtils';
import { consolidateDependsOnForLambda } from '../utils/consolidateDependsOn';
import { secretValuesWalkthrough } from './secretValuesWalkthrough';
import { secretNamesToSecretDeltas } from '../secrets/secretDeltaUtilities';
import { getLocalFunctionSecretNames } from '../secrets/functionSecretsStateManager';
import { tryUpdateTopLevelComment } from '../utils/updateTopLevelComment';
import { addLayersToFunctionWalkthrough } from './addLayerToFunctionWalkthrough';
import autogeneratedParameters from './autogeneratedParameters';
import { askExecRolePermissionsQuestions } from './execPermissionsWalkthrough';
import { generalQuestionsWalkthrough, settingsUpdateSelection } from './generalQuestionsWalkthrough';
import { scheduleWalkthrough } from './scheduleWalkthrough';

/**
 * Starting point for CLI walkthrough that generates a lambda function
 * @param context The Amplify Context object
 */
export async function createWalkthrough(
  context: $TSContext,
  templateParameters: Partial<FunctionParameters>,
): Promise<Partial<FunctionParameters>> {
  // merge in parameters that don't require any additional input
  templateParameters = merge(templateParameters, autogeneratedParameters(context));

  // ask generic function questions and merge in results
  if (!templateParameters.functionName) {
    templateParameters = merge(templateParameters, await generalQuestionsWalkthrough(context));
  }
  if (templateParameters.functionName) {
    templateParameters.resourceName = templateParameters.functionName;
  }

  // ask runtime selection questions and merge in results
  if (!templateParameters.runtime) {
    let runtimeSelection = await runtimeWalkthrough(context, templateParameters);
    templateParameters = merge(templateParameters, runtimeSelection[0]);
  }

  // ask template selection questions and merge in results
  templateParameters = merge(templateParameters, await templateWalkthrough(context, templateParameters));

  if (templateParameters.skipAdvancedSection) {
    return templateParameters;
  }

  // list out the advanced settings before asking whether to configure them
  context.print.info('');
  context.print.success('Available advanced settings:');
  advancedSettingsList.forEach(setting => context.print.info('- '.concat(setting)));
  context.print.info('');

  // ask whether to configure advanced settings
  if (await context.amplify.confirmPrompt('Do you want to configure advanced settings?', false)) {
    if (await context.amplify.confirmPrompt('Do you want to access other resources in this project from your Lambda function?')) {
      templateParameters = merge(
        templateParameters,
        await askExecRolePermissionsQuestions(context, templateParameters.functionName, undefined, templateParameters.environmentMap),
      );
    }

    // ask scheduling Lambda questions and merge in results
    templateParameters = merge(templateParameters, await scheduleWalkthrough(context, templateParameters));

    // ask lambda layer questions and merge in results
    templateParameters = merge(templateParameters, await addLayersToFunctionWalkthrough(context, templateParameters.runtime));

    // ask environment variable questions and merge in results
    if (await context.amplify.confirmPrompt('Do you want to configure environment variables for this function?', false)) {
      templateParameters = merge(templateParameters, await askEnvironmentVariableQuestions(templateParameters.functionName));
      //update top-level comment to be inserted into example
      templateParameters.topLevelComment = buildTopLevelComment( templateParameters.environmentMap );
      //show updated environment variables
      const envVarViewString = buildShowEnvVars(templateParameters.environmentMap);
      context.print.info(envVarViewString);
    }

    // ask function secrets questions and merge in results
    templateParameters = merge(
      templateParameters,
      await secretValuesWalkthrough(
        secretNamesToSecretDeltas(getLocalFunctionSecretNames(templateParameters.functionName)),
        Object.keys(getStoredEnvironmentVariables(templateParameters.functionName)),
      ),
    );
  }

  return templateParameters;
}

//Function to build message to be displayed when
//Lambda environment variables are enabled.
export function buildShowEnvVars(envVariableMap){
  const envVarViewArr = Object.keys(envVariableMap); //sorted list of existing env variables
  const envVarViewString= envVarViewArr.join("\n\t");
  const showEnvComment = `${envVarPrintoutPrefix}${envVarViewString}`;
  return showEnvComment;
}

//Insert 'Environment Map' keys - (env, region, resource-names, user supplied variables)
//into topLevelComment to be inserted into example files and displayed to the user
export function buildTopLevelComment( envVariableMap ){
  const envVarViewArr = Object.keys(envVariableMap);
  const envVarViewString= envVarViewArr.join("\n\t");
  const topLevelComment = `${topLevelCommentPrefix}${envVarViewString}${topLevelCommentSuffix}`;
  return topLevelComment;
}

function provideInformation(context, lambdaToUpdate, functionRuntime, currentParameters, scheduleParameters) {
  // Provide general information
  context.print.success('General information');
  context.print.info('- Name: '.concat(lambdaToUpdate));
  context.print.info('- Runtime: '.concat(functionRuntime));
  context.print.info('');

  // Provide resource access permission information
  context.print.success('Resource access permission');
  const currentCategoryPermissions = fetchPermissionCategories(currentParameters.permissions);
  if (currentCategoryPermissions.length) {
    currentCategoryPermissions.forEach(category => {
      const currentResources = fetchPermissionResourcesForCategory(currentParameters.permissions, category);
      currentResources.forEach(resource => {
        const currentPermissions = fetchPermissionsForResourceInCategory(currentParameters.permissions, category, resource);
        const formattedCurrentPermissions = ' ('.concat(currentPermissions.join(', ').concat(')'));
        context.print.info('- '.concat(resource).concat(formattedCurrentPermissions));
      });
    });
  } else {
    context.print.info('- Not configured');
  }
  context.print.info('');

  // Provide scheduling information
  context.print.success('Scheduled recurring invocation');
  if (scheduleParameters.cloudwatchRule && scheduleParameters.cloudwatchRule !== 'NONE') {
    context.print.info('- '.concat(scheduleParameters.cloudwatchRule));
    context.print.info('');
  } else {
    context.print.info('- Not configured');
    context.print.info('');
  }

  // Provide lambda layer information
  context.print.success('Lambda layers');
  if (currentParameters.lambdaLayers && currentParameters.lambdaLayers.length) {
    currentParameters.lambdaLayers.forEach(layer => {
      if (layer.arn) {
        context.print.info('- '.concat(layer.arn));
      } else {
        context.print.info(`- ${layer.resourceName}`);
      }
    });
    context.print.info('');
  } else {
    context.print.info('- Not configured');
    context.print.info('');
  }

  // Provide environment variables
  context.print.success('Environment variables:');
  const storedEnvironmentVariables = getStoredEnvironmentVariables(lambdaToUpdate);
  if (_.size(storedEnvironmentVariables) !== 0) {
    _.forEach(storedEnvironmentVariables, (environmentVariableValue, environmentVariableKey) => {
      context.print.info(`- ${environmentVariableKey}: ${environmentVariableValue}`);
    });
  } else {
    context.print.info('- Not configured');
  }
  context.print.info('');

  // secrets configuration
  context.print.success('Secrets configuration');
  const currentSecrets = getLocalFunctionSecretNames(lambdaToUpdate);
  if (currentSecrets.length) {
    currentSecrets.forEach(secretName => context.print.info(`- ${secretName}`));
  } else {
    context.print.info('- Not configured');
  }
  context.print.info('');
}

/**
 * TODO this function needs to be refactored so it doesn't have side-effects of writing to CFN files
 */
export async function updateWalkthrough(context: $TSContext, lambdaToUpdate?: string) {
  const lambdaFuncResourceNames = ((await context.amplify.getResourceStatus()).allResources as $TSAny[])
    .filter(resource => resource.service === ServiceName.LambdaFunction && resource.mobileHubMigrated !== true)
    .map(resource => resource.resourceName);

  if (lambdaFuncResourceNames.length === 0) {
    context.print.error('No Lambda function resource to update. Use "amplify add function" to create a new function.');
    return;
  }

  if (lambdaToUpdate) {
    if (!lambdaFuncResourceNames.includes(lambdaToUpdate)) {
      context.print.error(`No Lambda function named ${lambdaToUpdate} exists in the project.`);
      return;
    }
  } else {
    const resourceQuestion = [
      {
        name: 'resourceName',
        message: 'Select the Lambda function you want to update',
        type: 'list',
        choices: lambdaFuncResourceNames,
      },
    ];
    lambdaToUpdate = (await inquirer.prompt(resourceQuestion)).resourceName as string;
  }

  // initialize function parameters for update
  const functionParameters: Partial<FunctionParameters> = {
    resourceName: lambdaToUpdate,
    environmentMap: {
      ENV: {
        Ref: 'env',
      },
      REGION: {
        Ref: 'AWS::Region',
      },
    },
  };

  const projectBackendDirPath = pathManager.getBackendDirPath();
  const resourceDirPath = path.join(projectBackendDirPath, categoryName, functionParameters.resourceName);
  const currentParameters = loadFunctionParameters(resourceDirPath);
  const functionRuntime = context.amplify.readBreadcrumbs(categoryName, functionParameters.resourceName).functionRuntime as string;

  const cfnParameters = JSONUtilities.readJson<$TSAny>(path.join(resourceDirPath, parametersFileName), { throwIfNotExist: false }) || {};
  const scheduleParameters = {
    cloudwatchRule: cfnParameters.CloudWatchRule,
    resourceName: functionParameters.resourceName,
  };

  provideInformation(context, lambdaToUpdate, functionRuntime, currentParameters, scheduleParameters);

  // Determine which settings need to be updated
  const { selectedSettings }: $TSAny = await settingsUpdateSelection();

  if (selectedSettings.includes(resourceAccessSetting)) {
    const additionalParameters = await askExecRolePermissionsQuestions(context, lambdaToUpdate, currentParameters.permissions);
    additionalParameters.dependsOn = additionalParameters.dependsOn || [];
    merge(functionParameters, additionalParameters);
    updateCFNFileForResourcePermissions(resourceDirPath, functionParameters, currentParameters);
  }

  // ask scheduling Lambda questions and merge in results
  if (selectedSettings.includes(cronJobSetting)) {
    merge(functionParameters, await scheduleWalkthrough(context, scheduleParameters, true));
  }

  // ask lambdalayer questions and merge results
  if (selectedSettings.includes(lambdaLayerSetting)) {
    const currentFunctionParameters: $TSAny =
      JSONUtilities.readJson(path.join(resourceDirPath, functionParametersFileName), { throwIfNotExist: false }) || {};
    merge(
      functionParameters,
      await addLayersToFunctionWalkthrough(context, { value: functionRuntime }, currentFunctionParameters.lambdaLayers, true),
    );
    // writing to the CFN here because it's done above for the schedule and the permissions but we should really pull all of it into another function
    addLayerCFNParameters(context, functionParameters, resourceDirPath);
  }

  if (selectedSettings.includes(secretsConfiguration)) {
    merge(
      functionParameters,
      await secretValuesWalkthrough(
        secretNamesToSecretDeltas(getLocalFunctionSecretNames(functionParameters.resourceName)),
        Object.keys(getStoredEnvironmentVariables(functionParameters.resourceName)),
        {
          preConfirmed: true,
        },
      ),
    );
  }

  // consolidate dependsOn as above logic is overwriting
  const projectMeta = stateManager.getMeta();
  functionParameters.dependsOn = consolidateDependsOnForLambda(projectMeta, functionParameters.dependsOn, lambdaToUpdate, selectedSettings);
  merge(
    functionParameters,
    await askEnvironmentVariableQuestions(lambdaToUpdate, undefined, !selectedSettings.includes(environmentVariableSetting)),
  );
  //update top-level comment to be inserted into example
  functionParameters.topLevelComment = buildTopLevelComment( functionParameters.environmentMap );
  return functionParameters;
}

export function migrate(context: $TSContext, projectPath: string, resourceName: string) {
  const resourceDirPath = pathManager.getResourceDirectoryPath(projectPath, categoryName, resourceName);
  const cfnFilePath = path.join(resourceDirPath, `${resourceName}-cloudformation-template.json`);
  const oldCfn = JSONUtilities.readJson<$TSAny>(cfnFilePath);
  const newCfn: $TSAny = {};
  Object.assign(newCfn, oldCfn);

  // Add env parameter
  if (!newCfn.Parameters) {
    newCfn.Parameters = {};
  }
  newCfn.Parameters.env = {
    Type: 'String',
  };

  // Add conditions block
  if (!newCfn.Conditions) {
    newCfn.Conditions = {};
  }
  newCfn.Conditions.ShouldNotCreateEnvResources = {
    'Fn::Equals': [
      {
        Ref: 'env',
      },
      'NONE',
    ],
  };

  // Add if condition for resource name change
  const oldFunctionName = newCfn.Resources.LambdaFunction.Properties.FunctionName;

  newCfn.Resources.LambdaFunction.Properties.FunctionName = {
    'Fn::If': [
      'ShouldNotCreateEnvResources',
      oldFunctionName,
      {
        'Fn::Join': [
          '',
          [
            oldFunctionName,
            '-',
            {
              Ref: 'env',
            },
          ],
        ],
      },
    ],
  };

  newCfn.Resources.LambdaFunction.Properties.Environment = { Variables: { ENV: { Ref: 'env' } } };

  const oldRoleName = newCfn.Resources.LambdaExecutionRole.Properties.RoleName;

  newCfn.Resources.LambdaExecutionRole.Properties.RoleName = {
    'Fn::If': [
      'ShouldNotCreateEnvResources',
      oldRoleName,
      {
        'Fn::Join': [
          '',
          [
            oldRoleName,
            '-',
            {
              Ref: 'env',
            },
          ],
        ],
      },
    ],
  };

  JSONUtilities.writeJson(cfnFilePath, newCfn);
}

export function updateCFNFileForResourcePermissions(
  resourceDirPath: string,
  functionParameters: Partial<FunctionParameters>,
  currentParameters: $TSObject,
  apiResourceName?: string,
) {
  const cfnFileName = `${functionParameters.resourceName}-cloudformation-template.json`;
  const cfnFilePath = path.join(resourceDirPath, cfnFileName);
  const cfnContent = JSONUtilities.readJson<$TSAny>(cfnFilePath);
  const dependsOnParams = { env: { Type: 'String' } };

  Object.keys(functionParameters.environmentMap)
    .filter(key => key !== 'ENV')
    .filter(key => key !== 'REGION')
    .filter(resourceProperty => 'Ref' in functionParameters.environmentMap[resourceProperty])
    .forEach(resourceProperty => {
      dependsOnParams[functionParameters.environmentMap[resourceProperty].Ref] = {
        Type: 'String',
        Default: functionParameters.environmentMap[resourceProperty].Ref,
      };
    });

  cfnContent.Parameters = getNewCFNParameters(
    cfnContent.Parameters,
    currentParameters,
    dependsOnParams,
    functionParameters.mutableParametersState,
    apiResourceName,
  );

  if (!cfnContent.Resources.AmplifyResourcesPolicy) {
    cfnContent.Resources.AmplifyResourcesPolicy = {
      DependsOn: ['LambdaExecutionRole'],
      Type: 'AWS::IAM::Policy',
      Properties: {
        PolicyName: 'amplify-lambda-execution-policy',
        Roles: [
          {
            Ref: 'LambdaExecutionRole',
          },
        ],
        PolicyDocument: {
          Version: '2012-10-17',
          Statement: [],
        },
      },
    };
  }

  if (functionParameters.categoryPolicies.length === 0) {
    delete cfnContent.Resources.AmplifyResourcesPolicy;
  } else {
    cfnContent.Resources.AmplifyResourcesPolicy.Properties.PolicyDocument.Statement = functionParameters.categoryPolicies;
  }

  cfnContent.Resources.LambdaFunction.Properties.Environment.Variables = getNewCFNEnvVariables(
    cfnContent.Resources.LambdaFunction.Properties.Environment.Variables,
    currentParameters,
    functionParameters.environmentMap,
    functionParameters.mutableParametersState,
    apiResourceName,
  );

  JSONUtilities.writeJson(cfnFilePath, cfnContent);
  tryUpdateTopLevelComment(resourceDirPath, _.keys(functionParameters.environmentMap));
}

const addLayerCFNParameters = (context: $TSContext, functionParameters: Partial<FunctionParameters>, resourceDirPath: string) => {
  const cfnFileName = `${functionParameters.resourceName}-cloudformation-template.json`;
  const cfnFilePath = path.join(resourceDirPath, cfnFileName);
  const cfnContent = JSONUtilities.readJson<$TSAny>(cfnFilePath);
  functionParameters.lambdaLayers.forEach(layer => {
    const resourceName = _.get(layer as ProjectLayer, ['resourceName'], null);
    if (resourceName) {
      const param: string = `function${resourceName}Arn`;
      if (cfnContent.Parameters[param] === undefined) {
        cfnContent.Parameters[param] = {
          Type: 'String',
          Default: param,
        };
      }
    }
  });
  cfnContent.Resources.LambdaFunction.Properties.Layers = convertLambdaLayerMetaToLayerCFNArray(
    functionParameters.lambdaLayers,
    context.amplify.getEnvInfo().envName,
  );
  JSONUtilities.writeJson(cfnFilePath, cfnContent);
};
