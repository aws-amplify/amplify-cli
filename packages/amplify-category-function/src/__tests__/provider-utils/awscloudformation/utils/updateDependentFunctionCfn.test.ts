import { updateDependentFunctionsCfn } from '../../../../provider-utils/awscloudformation/utils/updateDependentFunctionCfn';
import { loadFunctionParameters } from '../../../../provider-utils/awscloudformation/utils/loadFunctionParameters';
import {
  getResourcesForCfn,
  generateEnvVariablesForCfn,
} from '../../../../provider-utils/awscloudformation/service-walkthroughs/execPermissionsWalkthrough';
import { $TSContext } from 'amplify-cli-core';
import { FunctionDependency } from 'amplify-function-plugin-interface/src';
import { updateCFNFileForResourcePermissions } from '../../../../provider-utils/awscloudformation/service-walkthroughs/lambda-walkthrough';
jest.mock('fs-extra');

jest.mock('../../../../provider-utils/awscloudformation/service-walkthroughs/execPermissionsWalkthrough');
jest.mock('../../../../provider-utils/awscloudformation/utils/loadFunctionParameters');
jest.mock('path');
jest.mock('../../../../provider-utils/awscloudformation/service-walkthroughs/lambda-walkthrough');

jest.mock('amplify-cli-core', () => ({
  JSONUtilities: {
    readJson: jest.fn(),
    writeJson: jest.fn(),
  },
}));
const contextStub = {
  amplify: {
    updateamplifyMetaAfterResourceUpdate: jest.fn(),
    copyBatch: jest.fn(),
  },
};

const allResources = [
  {
    build: true,
    providerPlugin: 'awscloudformation',
    service: 'Lambda',
    resourceName: 'fn2',
    dependsOn: [
      {
        category: 'api',
        resourceName: 'mock_api',
        attributes: ['GraphQLAPIIdOutput'],
      },
    ],
  },
  {
    build: true,
    providerPlugin: 'awscloudformation',
    service: 'Lambda',
    resourceName: 'fn3',
    dependsOn: [
      {
        category: 'api',
        resourceName: 'mock_api',
        attributes: ['GraphQLAPIIdOutput'],
      },
    ],
  },
];
const backendDir = 'randomPath';
const apiResourceName = 'mock_api';

const loadResourceParameters_mock = loadFunctionParameters as jest.MockedFunction<typeof loadFunctionParameters>;
const getResourcesforCFN_mock = getResourcesForCfn as jest.MockedFunction<typeof getResourcesForCfn>;
const generateEnvVariablesforCFN_mock = generateEnvVariablesForCfn as jest.MockedFunction<typeof generateEnvVariablesForCfn>;

const updateCFNFileForResourcePermissions_mock = updateCFNFileForResourcePermissions as jest.MockedFunction<
  typeof updateCFNFileForResourcePermissions
>;
const cfnResources = [{ resourceName: 'storageattr1@model(appsync)' }, { resourceName: 'storageattr2@model(appsync)' }];
const permissionPolicies = 'randomPermissionsforapiandstorage';
const dependsOn: FunctionDependency[] = [
  {
    category: 'api',
    resourceName: 'mock_api',
    attributes: ['GraphQLAPIIdOutput'],
  },
];
const environmentMap = {
  API_MOCK_API_GRAPHQLAPIIDOUTPUT: {
    Ref: 'apimock_apiGraphQLAPIIdOutput',
  },
};
const envVarStringList = '';

getResourcesforCFN_mock.mockResolvedValue({ permissionPolicies, cfnResources });
generateEnvVariablesforCFN_mock.mockResolvedValue({ dependsOn, environmentMap, envVarStringList });

test('update dependent functions', async () => {
  jest.clearAllMocks();
  const modelsDeleted = ['model3'];
  loadResourceParameters_mock
    .mockReturnValueOnce({
      permissions: {
        storage: {
          model1: ['create'],
          model2: ['create'],
          model3: ['create'],
        },
      },
    })
    .mockReturnValueOnce({
      permissions: {
        storage: {
          model3: ['create'],
        },
      },
    });
  await updateDependentFunctionsCfn(contextStub as unknown as $TSContext, allResources, backendDir, modelsDeleted, apiResourceName);
  expect(updateCFNFileForResourcePermissions_mock.mock.calls[0][1]).toMatchSnapshot();
});

test('update dependent functions', async () => {
  jest.clearAllMocks();
  const modelsDeleted = ['model1', 'model2'];
  loadResourceParameters_mock
    .mockReturnValueOnce({
      permissions: {
        storage: {
          model1: ['create'],
          model2: ['create'],
          model3: ['create'],
        },
      },
      lambdaLayers: [
        {
          type: 'ProjectLayer',
          resourceName: 'mocklayer',
          version: 1,
        },
      ],
    })
    .mockReturnValueOnce({
      permissions: {
        storage: {
          model3: ['create'],
        },
      },
      lambdaLayers: [
        {
          type: 'ProjectLayer',
          resourceName: 'mocklayer',
          version: 1,
        },
      ],
    });
  await updateDependentFunctionsCfn(contextStub as unknown as $TSContext, allResources, backendDir, modelsDeleted, apiResourceName);
  expect(updateCFNFileForResourcePermissions_mock.mock.calls[0][1]).toMatchSnapshot();
});
