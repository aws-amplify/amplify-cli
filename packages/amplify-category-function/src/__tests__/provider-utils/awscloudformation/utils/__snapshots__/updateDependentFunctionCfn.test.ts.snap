// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update dependent functions 1`] = `
Object {
  "categoryPolicies": Array [
    "randomPermissionsforapiandstorage",
    "randomPermissionsforapiandstorage",
  ],
  "dependsOn": Array [
    Object {
      "attributes": Array [
        "GraphQLAPIIdOutput",
      ],
      "category": "api",
      "resourceName": "mock_api",
    },
  ],
  "environmentMap": Object {
    "API_MOCK_API_GRAPHQLAPIIDOUTPUT": Object {
      "Ref": "apimock_apiGraphQLAPIIdOutput",
    },
  },
  "lambdaLayers": undefined,
  "mutableParametersState": Object {
    "permissions": Object {
      "storage": Object {
        "model1": Array [
          "create",
        ],
        "model2": Array [
          "create",
        ],
      },
    },
  },
  "resourceName": "fn2",
}
`;

exports[`update dependent functions 2`] = `
Object {
  "categoryPolicies": Array [
    "randomPermissionsforapiandstorage",
  ],
  "dependsOn": Array [
    Object {
      "attributes": Array [
        "GraphQLAPIIdOutput",
      ],
      "category": "api",
      "resourceName": "mock_api",
    },
  ],
  "environmentMap": Object {
    "API_MOCK_API_GRAPHQLAPIIDOUTPUT": Object {
      "Ref": "apimock_apiGraphQLAPIIdOutput",
    },
  },
  "lambdaLayers": Array [
    Object {
      "resourceName": "mocklayer",
      "type": "ProjectLayer",
      "version": 1,
    },
  ],
  "mutableParametersState": Object {
    "permissions": Object {
      "storage": Object {
        "model3": Array [
          "create",
        ],
      },
    },
  },
  "resourceName": "fn2",
}
`;
